service/s3/presign_post.go

```diff
@@ -0,0 +1,408 @@
+	PresignPost(
+		credentials aws.Credentials,
+		bucket string, key string,
+		region string, service string, signingTime time.Time, conditions []interface{}, expirationTime time.Time,
+		optFns ...func(*v4.SignerOptions),
+	) (fields map[string]interface{}, err error)
+}
+
+// PresignPostOptions represent the options to be passed to a PresignPost sign request
+type PresignPostOptions struct {
+
+	// ClientOptions are list of functional options to mutate client options used by
+	// the presign client.
+	ClientOptions []func(*Options)
+
+	// PostPresigner to use. One will be created if none is provided
+	PostPresigner PresignPost
+
+	// Expires sets the expiration duration for the generated presign url. This should
+	// be the duration in seconds the presigned URL should be considered valid for. If
+	// not set or set to zero, presign url would default to expire after 900 seconds.
+	Expires time.Duration
+
+	// Conditions a list of extra conditions to pass to the policy document
+	// Available conditions can be found [here]
+	//
+	// [here]https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTConstructPolicy.html#sigv4-PolicyConditions
+	Conditions []interface{}
+}
+
+type presignPostConverter PresignPostOptions
+
+// presignPostRequestMiddlewareOptions is the options for the presignPostRequestMiddleware middleware.
+type presignPostRequestMiddlewareOptions struct {
+	CredentialsProvider aws.CredentialsProvider
+	Presigner           PresignPost
+	LogSigning          bool
+	ExpiresIn           time.Duration
+	Conditions          []interface{}
+}
+
+type presignPostRequestMiddleware struct {
+	credentialsProvider aws.CredentialsProvider
+	presigner           PresignPost
+	logSigning          bool
+	expiresIn           time.Duration
+	conditions          []interface{}
+}
+
+// newPresignPostRequestMiddleware returns a new presignPostRequestMiddleware
+// initialized with the presigner.
+func newPresignPostRequestMiddleware(options presignPostRequestMiddlewareOptions) *presignPostRequestMiddleware {
+	return &presignPostRequestMiddleware{
+		credentialsProvider: options.CredentialsProvider,
+		presigner:           options.Presigner,
+		logSigning:          options.LogSigning,
+		expiresIn:           options.ExpiresIn,
+		conditions:          options.Conditions,
+	}
+}
+
+// ID provides the middleware ID.
+func (*presignPostRequestMiddleware) ID() string { return "PresignPostRequestMiddleware" }
+
+// HandleFinalize will take the provided input and create a presigned url for
+// the http request using the SigV4 presign authentication scheme.
+//
+// Since the signed request is not a valid HTTP request
+func (s *presignPostRequestMiddleware) HandleFinalize(
+	ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler,
+) (
+	out middleware.FinalizeOutput, metadata middleware.Metadata, err error,
+) {
+	if s.presigner == nil {
+		return out, metadata, fmt.Errorf("no presigner provided. Double-check your configuration")
```

Comment:
nit, can this actually happen? Since we're defaulting to one? I guess it could if they explicitly `nil` it out but not sure  if we care to guard for things like that. I think there are any number of things you could nil out on client config right now that would cause panics.

Reply:
Removed. 

This can't really happen "naturally", but I added it since during development I had a specific scenario where this was being set to nil, and it took me a while to figure it out.